core/src/main/java/io/grpc/transport/ClientTransport.java

```diff
@@ -69,6 +71,21 @@ ClientStream newStream(MethodDescriptor<?, ?> method,
   void start(Listener listener);
 
   /**
+   * Pings the remote endpoint to verify that the transport is still active. The future will
+   * complete when the remote endpoint replies. The future completes exceptionally if the transport
+   * is shutdown before the response is received. The value of the future is the round-trip time of
+   * the ping, in nanoseconds.
+   *
+   * <p>Transports that do not have any mechanism by which to ping the remote endpoint should
+   * return {@link com.google.common.util.concurrent.Futures#immediateFuture(Object)
+   * immediateFuture(0)}.
+   *
+   * @return a future that completes when the remote endpoint responds to the ping with a value of
+   *     the round-trip time in nanoseconds
+   */
+  ListenableFuture<Long> ping();
```

Comment:
@ejona86, I ended up having this return the round-trip time. For simplicity, I implemented it such that only a single ping is outstanding at any given moment. If a ping is request while one is already outstanding, it just gets back the result of the outstanding ping instead of sending another.

For that reason, measuring the round-trip time isn't so easy since there's no easy way, as a caller, to determine if you're piggy-backing on an outstanding ping or not. So the value of the ping is the round-trip time in nanoseconds.

I wonder if pulling in something like [this](https://github.com/jhump/bluegosling/blob/master/src/com/apriori/concurrent/Duration.java) would be useful as a way to represent the result, instead of just a boxed primitive.